### ISSUE TITLE
String valued enums differ from tsc

### ISSUE THREAD

**Issue by fiesh** [babel/babel#10785]
## Bug Report

I think this should have been fixed by #7160.

**Input Code**
```ts
declare const enum E {
        e0 = "e0",
}

function f()
{
        console.log(E.e0);
}
```

**Expected behavior/code**
This is `tsc`'s output:
```js
function f() {
    console.log("e0" /* e0 */);
}
```

This is babel's output using `@babel/preset-typescript`:
```js
function f() {
  console.log(E.e0);
}
```

- Babel version(s): v7.7.4
- Node/npm version: Node 12 / npm 6.12.0
- OS: Gentoo Linux
- Monorepo: Just test file
- How you are using Babel: `npm run babel -- --config-file /path/to/babelrc --out-file a.js a.ts` with `babelrc` looking like this:
```json
{
  "presets": [
    "@babel/preset-typescript"
  ]
}
```

**Comment by futagoza** [comment]
From [the website/docs](https://babeljs.io/docs/en/babel-plugin-transform-tyapescript#caveats) for this plugin:
> This plugin does not support const enums because those require type information to compile

It also says, as a workaround, to use [babel-plugin-const-enum](https://www.npmjs.com/package/babel-plugin-const-enum); or remove `const` which makes the enum available at runtime (which I'm presuming you don't want?).

**Comment by JLHwung** [comment]
Yeah, Babel should throw "const enums are not supported" on `declare const enum E`.

**Comment by JLHwung** [comment]
For a better Dev UX, we should throw an `const enums are not supported` error. Here is where the error is thrown, we should make sure it will also throw for `declare const enums`.

https://github.com/babel/babel/blob/a6e8b3843b53c6e106d9fb2e37fea60c306dd97c/packages/babel-plugin-transform-typescript/src/enum.js#L6-L14

If you don't know how to clone Babel, follow these steps: (you need to have `make` and `yarn` available on your machine).
1. Write a comment there to let other possible contributors know that you are working on this bug.
1. Fork the repo
1. Run `git clone https://github.com/<YOUR_USERNAME>/babel.git && cd babel`
1. Run `yarn && make bootstrap`
1. Wait :hourglass_flowing_sand: 
1. Run `make watch` (or `make build` whenever you change a file)
1. Add a test (only `input.js`; `options.json` will be automatically generated) to packages/babel-plugin-transform-typescript/test/fixtures/enum/declare-const
1. Update the code!
1. `yarn jest babel-plugin-transform-typescript` to run the tests
    - If some test outputs don't match but the new results are correct, you can delete the bad `options.json` files and run the tests again
1. If it is working, run `make test` to run all the tests
1. Run `git push` and open a PR!

**Comment by nicolo-ribaudo** [comment]
We did already throw that error, but we removed it due to a really high number of people requesting it

**Comment by nicolo-ribaudo** [comment]
Ok no, we currently throw 🤔
https://github.com/babel/babel/issues/6476

**Comment by nicolo-ribaudo** [comment]
Related: https://github.com/babel/babel/issues/6476

**Comment by fiesh** [comment]
> It also says, as a workaround, to use [babel-plugin-const-enum](https://www.npmjs.com/package/babel-plugin-const-enum); or remove `const` which makes the enum available at runtime (which I'm presuming you don't want?).

I used `babel-plugin-const-enum` but didn't see any difference, unless I used it incorrectly.

Also removing the `const` from the `enum` doesn't change anything, so I don't understand what it meant by not supporting `const enum`.  I guess a `const enum` is different from a `const` `enum` in a way?!

**Comment by fiesh** [comment]
OK, the problem is that `declare` removes the generated code when `const` is removed.

All our `enum`s are generated automatically in our (C++) backend.  We generate a declaration file that contains them, and since it's a declaration file, of course we can only write declarations.  So I guess a workaround would be to generate an additional implementation file that contains the `enum` implementations as non-`const`.  Hmm not ideal but doable.

**Comment by fiesh** [comment]
@JLHwung I also think this should only be a warning, not an error, at least in the `declare` case.  When one uses string `enum`s, one can also use the actual string value `"e0"` instead of `E.e0`, and TypeScript will still type check this.  Using strings as such should be legal and would work with babel I think.

**Comment by dosentmatter** [comment]
@fiesh, from what I can tell. It seems like babel currently transforms `declare const enum` as if it was `declare enum`. Unlike the plain `const enum`, `declare const enum` doesn't complain. Either way, babel doesn't support the inlining feature of const enums. The only way I know how to inline right now is by using [`babel-plugin-const-enum`](https://www.npmjs.com/package/babel-plugin-const-enum) with a minifier. I think the correct way for babel is the following:

Currently:
```
declare const enum E { // with const
  e0 = "e0",
}

function f()
{
  console.log(E.e0);
}

>becomes>

function f() {
  console.log(E.e0); // same as below (This is incorrect. TypeScript transpiler inlines. I propose babel should error just like how it does with plain `const enum`.
}
```

```
declare enum E { // without const
  e0 = "e0",
}

function f()
{
  console.log(E.e0);
}

>becomes>

function f() {
  console.log(E.e0); // same as above (this is correct though since TypeScript compiler transforms the same)
}
```

I am the creator of [`babel-plugin-const-enum`](https://www.npmjs.com/package/babel-plugin-const-enum). Currently, the `remove-const` transform works but it will give you the same result as what babel is currently doing, removing the `const`.

`const-object` should work but it is currently transforming like so:
```
declare const enum E { // with const
  e0 = "e0",
}

>becomes>

declare const E = { // invalid TS (A 'const' initializer in an ambient context must be a string or numeric literal or literal enum reference.(1254))
  e0: "e0"
}
```

If I just modify the transform to remove the `declare`, it should be functionally the same, after minification runs (which is the point of `const-object`). After minification, the generated object is inlined by terser/uglify/etc., so it's fine to have the intermediate const object in there. It should be simple and I can do it when I have time. Or you can submit a PR if you think it would be helpful.

**Comment by dosentmatter** [comment]
@fiesh, from what I can tell. It seems like babel currently transforms `declare const enum` as if it was `declare enum`. Unlike the plain `const enum`, `declare const enum` doesn't complain. Either way, babel doesn't support the inlining feature of const enums. The only way I know how to inline right now is by using [`babel-plugin-const-enum`](https://www.npmjs.com/package/babel-plugin-const-enum) with a minifier. I think the correct way for babel is the following:

Currently:
```
declare const enum E { // with const
  e0 = "e0",
}

function f()
{
  console.log(E.e0);
}

>becomes>

function f() {
  console.log(E.e0); // same as below (This is incorrect. TypeScript transpiler inlines. I propose babel should error just like how it does with plain `const enum`.
}
```

```
declare enum E { // without const
  e0 = "e0",
}

function f()
{
  console.log(E.e0);
}

>becomes>

function f() {
  console.log(E.e0); // same as above (this is correct though since TypeScript compiler transforms the same)
}
```

I am the creator of [`babel-plugin-const-enum`](https://www.npmjs.com/package/babel-plugin-const-enum). Currently, the `remove-const` transform works but it will give you the same result as what babel is currently doing, removing the `const`.

`const-object` should work but it is currently transforming like so:
```
declare const enum E { // with const
  e0 = "e0",
}

>becomes>

declare const E = { // invalid TS (A 'const' initializer in an ambient context must be a string or numeric literal or literal enum reference.(1254))
  e0: "e0"
}
```

If I just modify the transform to remove the `declare`, it should be functionally the same, after minification runs (which is the point of `const-object`). After minification, the generated object is inlined by terser/uglify/etc., so it's fine to have the intermediate const object in there. It should be simple and I can do it when I have time. Or you can submit a PR if you think it would be helpful.

A workaround, for now, is to remove your `declare` manually. Is there a reason you need the `declare`? `const enum`s are already removed from outputted code. `declare` also removes code. Adding the `declare` for `const enum` doesn't seem to do anything extra besides having semantic meaning that the enum may not have originated from the current TypeScript file.

**Comment by fiesh** [comment]
@dosentmatter Thank you very much for the detailed explanation.

Our use case is an automatically generated TypeScript declaration file that specifies the exposed API of the underlying C++ program.  Our enums were then exposed via `declare const enum`.  An enum `E` being `{ e0 = "e0" }` could then either be used via `E.e0` or the string literal `"e0"`.

The latter should probably be supported by babel since the `enum` can just be removed entirely.

Now we go an entirely different way and avoid TypeScript's soso enums altogether:

```ts
declare type E = "e0" | "e1";
```

And then, in an implementation file, create a literal object that's named like the type:

```ts
export const E = {
    e0: "e0",
    [Symbol.iterator]: function(): IterableIterator<FileType> { return [this.e0, ][Symbol.iterator](); }
} as const;
```

This allows the syntax `E.e0` and, on top of that, iterating over `E`.

So basically, we learned our lessen and just avoid TypeScript's enums altogether ;-)

**Comment by dosentmatter** [comment]
@fiesh, I think babel-plugin-typescript doesn't support inlining `const enum` because it requires type information and babel transpiles files/modules in isolation (`isolatedModules` flag). So if you do `export const enum`, that `const enum` won't be able to make it into a different module when importing it. `const enum` is entirely a type, since it inlines and has no code output.

The way `babel-plugin-const-enum` works around this is to transpile all `const enum`s into simple object literals that can further be inlined with a minifier.

Regarding your current implementation:
```
export const E = {
    e0: "e0",
    [Symbol.iterator]: function(): IterableIterator<FileType> { return [this.e0, ][Symbol.iterator](); }
} as const;
```
Can the iterator part be simplified to the following to avoid creating an extra function?
```
[Symbol.iterator]: [this.e0][Symbol.iterator]
```
I guess this would hog memory though if you have many properties, since the array always exists. The function way would allocate an array every call.

In the code you have, everything besides the `[Symbol.iterator]` is what `babel-plugin-const-enum` would transpile a `const enum` into. Here is a better [example](https://github.com/babel/babel/issues/8741#issuecomment-509041135).

When you feed the output into terser, you get:
![image](https://user-images.githubusercontent.com/5732253/70914561-176d0980-1fcd-11ea-9902-f08a29aa236c.png)

Note that you do get a tiny bit more output by adding in the iterator. So if you don't need the iterator, `babel-plugin-const-enum` might still be useful for you to type a little less boilerplate and a little smaller bundle size.

**Comment by sosukesuzuki** [comment]
How is the status? Can I work on this?

**Comment by dosentmatter** [comment]
> How is the status? Can I work on this?

@sosukesuzuki, I worked on my plugin, `babel-plugin-const-enum` https://github.com/dosentmatter/babel-plugin-const-enum/issues/16, yesterday and it turns out I already support `declare const enum`. I released `babel-plugin-const-enum@1.0.0` already. It will just ignore the `declare` for `const enum` and transpile with `removeConst` or `constObject`, depending on which option you select. `constObject` with a minifier will get you the same result as the TypeScript Compiler - see https://github.com/babel/babel/issues/8741.

On the `@babel/plugin-transform-typescript` side of things:
1. `declare enum` transpilation output is correct. The enum gets removed and access gets transpiled to `E.eo`, which is the same as `tsc`.
2. `declare const enum` currently transpiles as if it was `declare enum`. I think this should be changed to throw an error `'const' enums are not supported.` to match the error for `const enum`.

If you want to work on this, I think the only thing to do is to submit a PR to `@babel/plugin-transform-typescript` to fix 2. - throw an error when it sees `declare const enum`.

[`const enum` repl error](https://babeljs.io/repl#?browsers=&build=&builtIns=false&spec=false&loose=false&code_lz=MYewdgzgLgBApmArgWxgUSag3gKAL5A&debug=false&forceAllTransforms=false&shippedProposals=false&circleciRepo=&evaluate=false&fileSize=false&timeTravel=false&sourceType=module&lineWrap=true&presets=env%2Ctypescript%2Cenv&prettier=false&targets=&version=7.9.0&externalPlugins=)
![image](https://user-images.githubusercontent.com/5732253/79076636-8d9a2800-7cb0-11ea-88ea-618441e4cc79.png)

[`declare const enum` repl no error](https://babeljs.io/repl#?browsers=&build=&builtIns=false&spec=false&loose=false&code_lz=CYUwxgNghgTiAEYD2A7AzgF3iFBXAtvAKJ6EDeAUAL5A&debug=false&forceAllTransforms=false&shippedProposals=false&circleciRepo=&evaluate=false&fileSize=false&timeTravel=false&sourceType=module&lineWrap=true&presets=env%2Ctypescript%2Cenv&prettier=false&targets=&version=7.9.0&externalPlugins=)
![image](https://user-images.githubusercontent.com/5732253/79076651-ba4e3f80-7cb0-11ea-992a-f733bacbfdb6.png)
